### PR TITLE
Implement the Composition API for RadChart

### PR DESCRIPTION
**Pull request by APopatanasov**
- Axis;
- Ticks;
- CatesianChartGrid;
- ScatterLineSeries,
- BarSeries;
- CandleStickSeries;
- OhlcSeries;
- Trend Line and Financial Indicator;

Added ContainerVisualsFactoruy class that provides the ability to choose whether the visual elements should be drawn using the Composition API, to change the ContainerVisual used for the drawing of the element (currently only SpriteVisual could be used) and to modify the ContainerVisual by changing its Offset, Size, Brush etc.

Implemented a LabelCreator interface tha provides the ability using a bool method to indicate whether a Label should be created or not - that provides the ability to choose the exact Labels the user could create based on some implemetned logic and could improve the performace a lot for specific scenaries where only a few specific labels needs to be visualized.

### CODE REVIEW

File: Controls/Chart/Chart.UWP/Engine/Axes/AxisModel.cs [hunk 1]
```diff
@@ -102,6 +103,21 @@ public IContentFormatter ContentFormatter
         }
 
         /// <summary>
+        /// Gets or sets the <see cref="ILabelCreator"/> instance used to decide whether Label should be created.
+        /// </summary>
+        public ILabelCreator LabelCreator
```
Comment (TKonov) [inline Controls/Chart/Chart.UWP/Engine/Axes/AxisModel.cs]: Don't we use Factory as unified term

File: Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs [hunk 2]
```diff
@@ -532,11 +569,20 @@ private int ArrangeLines()
                         AxisTickModel nextTick = stripe.AssociatedTick.NextMajorTick;
                         if (nextTick.isVisible)
                         {
-                            line = this.GetLineVisual(lineCount);
                             RadRect lastRect = this.Orientation == Windows.UI.Xaml.Controls.Orientation.Vertical ?
-                                new RadRect(stripe.BorderRect.Right, stripe.BorderRect.Y, 1, stripe.BorderRect.Height) :
-                                new RadRect(stripe.BorderRect.X, stripe.BorderRect.Y - 1, stripe.BorderRect.Width, 1);
-                            this.ArrangeLine(lastRect, line);
+                                    new RadRect(stripe.BorderRect.Right, stripe.BorderRect.Y, 1, stripe.BorderRect.Height) :
```
Comment (TKonov) [inline Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs]: Can we wire this hardcoded 1 to a property, style or at least constant

File: Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+            }
+
+            if (dashArray[0] > 0 && dashArray[1] == 0)
+            {
+                if (containerVisual.Children.Count > 0)
+                {
+                    containerVisual.Children.RemoveAll();
+                }
+
+                this.SetCompositionColorBrush(containerVisual, new SolidColorBrush(Color.FromArgb(0x23, 0, 0, 0)));
```
Comment (TKonov) [inline Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs]: can we use some resource or at least named constant

File: Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+                    return this.CreateCandleStick(containerVisual, ohlcDataPoint);
+                }
+
+                if (dataPoint.Presenter is OhlcSeries)
+                {
+                    return this.CreateOhlcStick(containerVisual, ohlcDataPoint);
+                }
+            }
+
+            this.SetCompositionColorBrush(containerVisual, new SolidColorBrush(Color.FromArgb(0xFF, 0x1E, 0x98, 0xE4)));
```
Comment (TKonov) [inline Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs]: can we use some resource or at least named constant

File: Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs [hunk 5]
```diff
@@ -469,6 +480,23 @@ public void Update()
                 }
             }
 
+            public void UpdateVisuals()
+            {
+                if (dashArray == null)
+                {
+                    this.dashArray = new DoubleCollection() { 4, 2};
```
Comment (TKonov) [inline Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs]: can we take this from style or constant

Comment (TKonov) [inline Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs]: can the user override this

Comment (APopatanasov) [inline Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs]: We can take it from the Style. The reason I didn't do it was to distinguish the Composition from the Styles used by the visual elements that draw the ChartGrid when Composition is not used. 

It couldn't be overridden, but I have provided an easy way for the users to set the desired dash array through the  PrepareCartesianChartGridLineVisual method - there they can provide any desired dash array.

File: Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs (hunk 2 above)
Comment (APopatanasov) [inline Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs]: Yes we can. I believe this the width/height of the Line visualized in the ChartGrid.

File: Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs (hunk 3 above)
Comment (APopatanasov) [inline Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs]: I was not able to find an easy way how to get the exact same resource used by the visual elements. Also, for some values a named brushes are used while others are directly set to the Setters of the Styles to the Background/Fill/Foreground properties. 

Constants according to me are more appropriate - there are only 2-3 brushes that are used.

File: Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs (hunk 4 above)
Comment (APopatanasov) [inline Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs]: The same answer as the above one. 😄

File: Controls/Chart/Chart.UWP/Engine/Axes/AxisModel.cs (hunk 1 above)
Comment (APopatanasov) [inline Controls/Chart/Chart.UWP/Engine/Axes/AxisModel.cs]: I tried to put it into the ChartSeriesLabelDefinition class. Event though the class is public  it is initialized and used only internally and couldn't think of any other easy way how to users could implement their logic about which label should be created. 

I believe the approach of introducing such interface where they can implement their logic is much more straightforward for the users.

About the name as SpriteVisual instances are created I named it factory. Any suggestions about another name is appreciated. 😄

Comment (TKonov) [inline Controls/Chart/Chart.UWP/Engine/Axes/AxisModel.cs]: let's leave it like that then

File: Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs (hunk 5 above)
Comment (TKonov) [inline Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs]: great

File: Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs (hunk 2 above)
Comment (TKonov) [inline Controls/Chart/Chart.UWP/Visualization/CartesianChart/CartesianChartGrid.cs]: so, let's move it constant

File: Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs (hunk 3 above)
Comment (TKonov) [inline Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs]: Let's go with constants.

File: Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs (hunk 4 above)
Comment (TKonov) [inline Controls/Chart/Chart.UWP/Visualization/Common/ContainerVisualsFactory.cs]: lets do it

**Pull request merged**